guacamole/src/main/java/org/apache/guacamole/rest/connection/ConnectionResource.java

```diff
@@ -150,18 +153,26 @@ public ConnectionResource(@Assisted UserContext userContext,
      * @throws GuacamoleException
      *     If an error occurs while retrieving the connection history.
      */
-    @GET
+    @SuppressWarnings("deprecation")
     @Path("history")
-    public List<APIConnectionRecord> getConnectionHistory()
+    public ConnectionHistoryResource getConnectionHistory()
             throws GuacamoleException {
 
-        // Retrieve the requested connection's history
-        List<APIConnectionRecord> apiRecords = new ArrayList<APIConnectionRecord>();
-        for (ConnectionRecord record : connection.getHistory())
-            apiRecords.add(new APIConnectionRecord(record));
-
-        // Return the converted history
-        return apiRecords;
+        try {
+            return new ConnectionHistoryResource(connection.getConnectionHistory());
+        }
+        catch (GuacamoleUnsupportedException e) {
+            try {
+                List<ConnectionRecord> history = Collections.emptyList();
+                for (ConnectionRecord record : connection.getHistory()) {
+                    history.add(record);
+                }
+                return new ConnectionHistoryResource(new SimpleActivityRecordSet<>(history));
+            }
+            catch (GuacamoleUnsupportedException ex) {
+                return new ConnectionHistoryResource(new SimpleActivityRecordSet<>());
+            }
+        }
```

Comment:
This is the area that I'm talking about in terms of the conversion.  So, what I initially tried to do was:
```
try {
    return new ConnectionHistoryResource(connection.getConnectionHistory());
}
catch (GuacamoleUnsupportedException e) {
    try {
        return new ConnectionHistoryResource(new SimpleActivityRecordSet<ConnectionRecord>(connection.getHistory()));
    }
    catch (GuacamoleUnsupportedException ex) {
        return new ConnectionHistoryResource(new SimpleActivityRecordSet<>());
    }
}
```
However, when I do that, I get a compile error:

    incompatible types: List<CAP#1> cannot be converted to List<ConnectionRecord>

This appears to be because the return type of `getHistory()` is `List<? extends ConnectionRecord>` whereas the input type of `SimpleActivityRecordSet()` is `ActivityRecord`.  For some reason that `? extends` causes it a bunch of heartache.

Reply:
Ah, I see what you're saying. It's a rather confusing aspect of Java generics. Consider if this was possible:

```java
List<FancyThing> a = new ArrayList<>();
List<? extends Thing> b = a; // This is completely OK
List<Thing> c = b; // This shouldn't be allowed because ...
c.add(new Thing()); // ... this breaks the contract of List<FancyThing>
```

Suddenly, that `List` that's expected to contain only fancy things contains a non-fancy thing, breaking the contract of the type. Declaring something as `List<Thing>` is saying "a list of `Thing` or any subclass of `Thing`". Declaring something as `List<? extends Thing>` is saying "a list of **an unspecified subclass of `Thing`**".

For the approach here with `SimpleActivityRecordSet`:

* There is a convenient way to turn `List<? extends T>` into `List<T>`, [`Collections.unmodifiableList()`](https://docs.oracle.com/javase/8/docs/api/java/util/Collections.html#unmodifiableList-java.util.List-).
* If you're adding a convenience constructor to `SimpleActivityRecordSet()` that accepts a list of records, you should probably use `List<? extends RecordType>` for the parameter so that it can accept any `List` having usable elements. This is what `ArrayList` does, presumably for the same reason: https://docs.oracle.com/javase/8/docs/api/java/util/ArrayList.html#ArrayList-java.util.Collection-

Reply:
...unless the intent is that `SimpleActivityRecordSet` will be backed by the provided `List`, in which case `List<RecordType>` **is** correct. This is what is done with `SimpleDirectory`: https://guacamole.apache.org/doc/guacamole-ext/org/apache/guacamole/net/auth/simple/SimpleDirectory.html#SimpleDirectory-java.util.Collection-

I would suggest that the type be `Collection<RecordType>`, though, for API's sake.